### ISSUE TITLE
Architecture not fully consistent with article

### ISSUE THREAD

**Issue by gandroz** [lucidrains/conformer#5]
Hi
I think there are two smal mistakes ni your implementation.

1- The feed forward module starts with a normalization layer which is omitted in your code. Furthermore, the module is more the module as your code it plus the identity. That's done in the conformer block, but maybe is can be more consistent to do it directly in the FFN ?
2- In the conformer block, the article mentionned that the MHSA is sandwitched between two half-step feed-forward modules, not full modules

**Comment by lucidrains** [comment]
@gandroz Hello! For #2, you can see that is what I have done here https://github.com/lucidrains/conformer/blob/master/conformer/conformer.py#L190-L191

For #1, the `PreNorm` class takes care of normalization before executing the branch code

**Comment by lucidrains** [comment]
I don't really get the last part of your first question, can you expand on that?

**Comment by lucidrains** [comment]
@gandroz Perhaps a pull request would make it more clear for me and everyone else what you mean

**Comment by gandroz** [comment]
Sorry for Question2, I did not notice the scaling :(

For Question1, instead of using a `PreNorm`, why don't you simply do something like

```
self.net = nn.Sequential(
       nn.LayerNorm(dim),  # normalize here
       nn.Linear(dim, dim * mult),
       Swish(),
       nn.Dropout(dropout),
       nn.Linear(dim * mult, dim),
       nn.Dropout(dropout)
   )
```

and never mind for the last part of the question, the way you did it is exactly as it is done in the article. What I proposed was simply to do something like

```
class FeedForward(nn.Module):
    ....
    def forward(self, x):
        return x + 0.5 * self.net(x)
```

and thus

```
class ConformerBlock(nn.Module):
    ....
    def forward(self, x, mask = None):
        x = self.ff1(x)
        x = self.attn(x, mask = mask)  # with the same idea
        x = self.conv(x)  # with the same idea
        x = self.ff2(x)  # with the same idea
        x = self.post_norm(x)
        return x
```

**Comment by lucidrains** [comment]
It's just a matter of style, because in some transformers architectures, you can also toggle it to be post normalization, so I like to keep that logic outside of the branch module.
